cadc-util/src/main/java/ca/nrc/cadc/auth/SSLUtil.java

```diff
@@ -618,35 +547,73 @@ public static X509CertificateChain readPemCertificateAndKey(File pemFile)
+        InputStreamReader r = new InputStreamReader(new ByteArrayInputStream(data));
+        PEMParser parser = new PEMParser(r);
+        return readPEM(parser);
+    }
+    
+    private static X509CertificateChain readPEM(PEMParser parser) 
+        throws InvalidKeySpecException, NoSuchAlgorithmException, IOException, CertificateException {
+        
+        PrivateKey privateKey = null;
+        List<byte[]> certs = new ArrayList<>();
+        int byteSize = 0;
 
-        return new X509CertificateChain(chain, pk);
+        Object obj = parser.readObject();
+        while (obj != null) {
+            log.debug("found: " + obj.getClass().getName());
+            // if private key first: PEMKeyPair followed by PemObject(s) with type=certificate
+            // if cert/key/cert...: X509CertificateHolder followed by PemObject(s)
+            if (obj instanceof PEMKeyPair) {
+                PEMKeyPair pkp = (PEMKeyPair) obj;
+                PrivateKeyInfo pki = pkp.getPrivateKeyInfo();
+                JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+                privateKey = converter.getPrivateKey(pki);
+                log.debug(" private key: " + privateKey.getEncoded().length);
+            } else if (obj instanceof PrivateKeyInfo) {
+                PrivateKeyInfo pki = (PrivateKeyInfo) obj;
+                JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+                privateKey = converter.getPrivateKey(pki);
+                log.debug(" private key: " + privateKey.getEncoded().length);
+            } else if (obj instanceof X509CertificateHolder) {
+                X509CertificateHolder xch = (X509CertificateHolder) obj;
+                byte[] bytes = xch.toASN1Structure().getEncoded("DER");
+                certs.add(bytes);
+                byteSize += bytes.length;
+                log.debug(" certificate: " + bytes.length);
+            } else if (obj instanceof PemObject) {
+                PemObject po = (PemObject) obj;
+                if ("certificate".equalsIgnoreCase(po.getType())) {
+                    certs.add(po.getContent());
+                    byteSize += po.getContent().length;
+                    log.debug(" certificate: " + po.getContent().length);
+                } else if ("rsa private key".equalsIgnoreCase(po.getType())) {
+                    KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+                    PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(po.getContent());
+                    privateKey = keyFactory.generatePrivate(keySpec);
+                    log.debug(" private key: " + po.getContent().length);
+                } else {
+                    log.warn("readPEM: unexpected PemObject type: " + po.getType() + " aka " + obj.getClass().getName());
+                }
+            }
+            
+            obj = parser.readPemObject();
+        }
+        
```

Comment:
Should it check here if the private key was found?

Reply:
no, it should be able to read a file with only certificates - we don't do it but it is a thing in some 
contexts